pkg/k8s/service_cache.go

```diff
@@ -420,6 +426,54 @@ func (s *ServiceCache) UniqueServiceFrontends() FrontendList {
 	return uniqueFrontends
 }
 
+// filterEndpoints filters local endpoints by using k8s service heuristics.
+// For now it only implements the topology aware hints.
+func (s *ServiceCache) filterEndpoints(localEndpoints *Endpoints, svc *Service) *Endpoints {
+	if !option.Config.EnableServiceTopology || svc == nil || !svc.TopologyAware {
+		return localEndpoints
+	}
+
+	if s.selfNodeZoneLabel == "" {
+		// The node doesn't have the zone label set, so we cannot filter endpoints
+		// by zone. Therefore, return all endpoints.
+		return localEndpoints
+	}
+
+	if svc.TrafficPolicy == loadbalancer.SVCTrafficPolicyLocal {
+		// According to https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/#constraints:
+		// """
+		// Topology Aware Hints are not used when either externalTrafficPolicy or
+		// internalTrafficPolicy is set to Local on a Service.
+		// """
+		return localEndpoints
+	}
+
+	filteredEndpoints := &Endpoints{Backends: map[string]*Backend{}}
+
+	for key, backend := range localEndpoints.Backends {
+		if len(backend.HintsForZones) == 0 {
+			return localEndpoints
+		}
+
+		zoneMatch := false
+		for _, hint := range backend.HintsForZones {
+			if hint == s.selfNodeZoneLabel {
+				zoneMatch = true
+				break
+			}
+		}
+		if zoneMatch {
+			filteredEndpoints.Backends[key] = backend
+		}
```

Comment:
`zoneMatch` is unnecessary. 

```suggestion
		for _, hint := range backend.HintsForZones {
			if hint == s.selfNodeZoneLabel {
		                filteredEndpoints.Backends[key] = backend
				break
			}
		}
```